### ISSUE TITLE
Allow re-installation from local remotes

### ISSUE THREAD

**Issue by smingerson** [rstudio/renv#284]
I am using drake in tandem with renv. I've decided to wrap the meat of my project in an R package, exposed to drake through `expose_imports()`. When changes require updating the package, I would expect `renv::install(mypkg)` to work, as long as I had bumped the package version. Instead, I get
`Error: package 'mypkg' is not available`. If I use `renv::install(dir/to/mypkg)`, everything works as expected. 

What's weird about this behavior is based on the renv.lock contents, I'm not sure why the first doesn't work.
```
    "mypkg": {
      "Package": "mypkg",
      "Version": "0.1.1",
      "Source": "Local",
      "RemoteType": "local",
      "RemoteUrl": "S:\\Projects\\mypkg",
      "Hash": "c42f78862cbba49422e3e92bf84eeeb4"
    }
```
I've also set RENV_PATHS_LOCAL to "S://Projects" or "file://S://Projects" in an attempt to fix it.

edit: I've now realized that `renv::update()` is more appropriate, given the package is already installed in the project. But... 
```
renv::update('mypkg')
# * Checking for updated packages ... Error in names(updates) <- extract_chr(updates, "Package") : 
#  attempt to set an attribute on NULL
```

One thing I noticed is `renv_snapshot_r_packages()` is returning `mypkg$Source` as "unknown", rather than "Local", so it isn't picking up that "RemoteURL" in the renv.lock as I expected.

I'm on the latest dev version, and had the same issues with the version on CRAN.

Another edit: I see you've addressed essentially this at #250. I agree with the conclusion there, but I find the behavior quite surprising. At the very least, a more informative error message or an explanation in [this vignette](https://rstudio.github.io/renv/articles/local-sources.html) about this behavior would be very welcome. I'd be happy to submit a PR for whichever is the desired route.

**Comment by kevinushey** [comment]
> When changes require updating the package, I would expect renv::install(mypkg) to work, as long as I had bumped the package version.

Currently, the function:

```
renv::install("package")
```

will only attempt to install the package from CRAN -- the locals system does not come into play (it currently only does for restore).

> One thing I noticed is renv_snapshot_r_packages() is returning mypkg$Source as "unknown", rather than "Local", so it isn't picking up that "RemoteURL" in the renv.lock as I expected.

Perhaps I'm misunderstanding, but `renv::snapshot()` is used to _generate_ the lockfile `renv.lock`; it does not attempt to read any information from `renv.lock`. The information it gleans is based on the DESCRIPTION files of packages installed in your project library.

Either way, `renv` will infer a local source if a local source is available, as per here:

https://github.com/rstudio/renv/blob/4f5f7e83fab8947b0ea2139372ba1385e766ddb4/R/snapshot.R#L613-L615

I think ultimately the feature request here is that `renv::install("pkgName")` should also look within the locals directories, rather than assuming the package must be available on CRAN.

**Comment by smingerson** [comment]
No misunderstanding on your end. Thank you for the clarifications. I think the crux of my question is "What is the 'right' way to update local packages within renv?". I'm hoping the answer is not "Set up an internal CRAN-like repository".

Your suggestion above is what I was requesting, but perhaps that is overzealous. It would be unfortunate for the same package to exist in both places and cause ambiguity.  Maybe `install()` requires a fully qualified path, but `update()` looks in `RENV_PATHS_LOCAL` if source is Local?

**Comment by kevinushey** [comment]
> I think the crux of my question is "What is the 'right' way to update local packages within renv?". I'm hoping the answer is not "Set up an internal CRAN-like repository".

I agree this is a fairly large overhead, but note that you are ultimately creating something like an "ad-hoc" local CRAN repository, since any restores will require the requisite version(s) of your local packages to be available as well. And, depending on how you plan to use `renv`, it could become more difficult to restore these packages on different machines, or in different environments.

Basically, the locals mechanism is useful as a crutch, but be careful on taking it too far -- it became the cause of a number of headaches in Packrat, for example.

> Maybe install() requires a fully qualified path, but update() looks in RENV_PATHS_LOCAL if source is Local?

This sounds reasonable to me. I'm less sure what the right behavior for `renv::install()` should be, but `renv::update()` should definitely respect the locals system, especially for packages with a Local source.

**Comment by smingerson** [comment]
Agree with you on the similarity between what I'm doing and a CRAN-like repository. 

The ultimate home for this package is Azure DevOps. The situation I described above is simply to make development more convenient and not requiring committing/pushing every single change. This could also be remedied by moving my drake plan into a package vignette and using `devtools::load_all()`, but since upon completion I want them separated, I started with them separated.

I don't have strong opinions on what the right behavior for `renv::install()` is, but from reading the documentation it wasn't clear to me that what I tried was not supported. 

On another note, once I have "mypkg" binaries saved in `RENV_PATHS_LOCAL`, then run `install(path/to/mypkg)`, I get a note that it won't be installed from `path/to/mypkg`, and instead use the binaries in `RENV_PATHS_LOCAL`. This partial adherence to using `RENV_PATHS_LOCAL` was unexpected.

**Comment by kevinushey** [comment]
Fair enough, it sounds like `RENV_PATHS_LOCAL` is the quickest way forward then.

> On another note, once I have "mypkg" binaries saved in RENV_PATHS_LOCAL, then run install(path/to/mypkg), I get a note that it won't be installed from path/to/mypkg, and instead use the binaries in RENV_PATHS_LOCAL. This partial adherence to using RENV_PATHS_LOCAL was unexpected.

Are you using the latest version of `renv`? This should be fixed in https://github.com/rstudio/renv/commit/8a60ecd493b5478377f91809bb39b2b963e7e635.

**Comment by smingerson** [comment]
I believe I am: 

```
  "renv": {
      "Package": "renv",
      "Version": "0.8.3-66",
      "Source": "GitHub",
      "RemoteType": "github",
      "RemoteHost": "api.github.com",
      "RemoteUsername": "rstudio",
      "RemoteRepo": "renv",
      "RemoteRef": "master",
      "RemoteSha": "c93cede1e455b4769a9c253de9d6914355e672e6",
      "Hash": "e03665c78ec6027276ee718af9e9e891"
    },
```
`renv_snapshot_r_packages()` within `renv::install(S:/Projects/mypkg)` gives a Source of "unknown", 
while `renv_retrieve(packages)` gives a source of "Local". Meanwhile, my lockfile also says unknown, but does have a remoteURL recorded. Let me know if other information would be helpful. Thank you for your responsiveness and clear explanations!

**Comment by smingerson** [comment]
I realized I can use `devtools::load_all(path/to/pkg)` in `drake::make`'s prework argument instead of going through package build & `renv::install()`. Talk about going about doing things the hard way.

**Comment by kevinushey** [comment]
>       "Version": "0.8.3-66",

This was fixed in 0.8.3-70, so you are just a little bit too out-of-date.

> I realized I can use devtools::load_all(path/to/pkg) in drake::make's prework argument instead of going through package build & renv::install().

This sounds like a much more natural approach :-)

**Comment by kevinushey** [comment]
I think the issue here is now ultimately well-understood, so I think this can be closed. Feel free to re-open or file another issue if you're still having issues.
